### DISCUSSION TITLE
pdgen remove stripes with [WARNING PDN-0200] Removing floating shape:

### DISCUSSION THREAD

**Discussion by titan73** [The-OpenROAD-Project/OpenROAD#5333] [Q&A] [answered]
Based on the example at the bottom of https://openroad.readthedocs.io/en/latest/main/src/pdn/README.html.

I did the following floorplan with the .def containing the power pins:

    read_def -floorplan_initialize floor.def
    set fp_area [ord::get_die_area]
    initialize_floorplan -die_area $fp_area -core_area $fp_area -site CORECUBRICK

    add_global_connection -net VDDD -pin_pattern {^VDDD$} -power
    add_global_connection -net GNDD -pin_pattern {^GNDD$} -ground

    set_voltage_domain -power VDDD -ground GNDD

    define_pdn_grid -name "Core"
    add_pdn_ring -grid "Core" -layers {metal4B alucap} -widths 5.0 -spacings 2.0 -core_offsets 4.5 -connect_to_pads

    add_pdn_stripe -followpins -layer metal1 -extend_to_core_ring

    add_pdn_stripe -layer metal4B -width 0.48 -pitch 56.0 -offset 2.0 -extend_to_core_ring
    add_pdn_stripe -layer alucap -width 1.40 -pitch 40.0 -offset 2.0 -extend_to_core_ring

    add_pdn_connect -layers {metal4B alucap}
    add_pdn_connect -layers {alucap metal4B}

    pdngen

The sroutes and the ring are created but not the stripes and I get the the warnings:

    [WARNING PDN-0200] Removing floating shape: GNDD (1.7600, -9.5000) - (2.2400, 821.3000) on metal4B
    [WARNING PDN-0200] Removing floating shape: VDDD (29.7600, -16.5000) - (30.2400, 828.3000) on metal4B

Is there something more to do to connect these shapes?

**Comment by titan73** [comment]
Also note that the related power pins imported the from .def are removed.

**Comment by maliberty** [comment]
Can you provide a test case?  Its hard to say much from just the output

**Reply by maliberty** [reply]
If so you should open an issue.

**Reply by titan73** [reply]
Yes. I'll make one tomorrow.

**Reply by titan73** [reply]
Are there usable sets of lefs (technology, sites and standard cells) to make a test case? sky130 or gf180. 
There are a lot of lefs in OR and ORFS gits but I'm unable to find a good set.

**Reply by titan73** [reply]
I make a test case with GF180 but I don't have the issue

    read_lef OpenROAD/src/tap/test/gf180/gf180mcu_5LM_1TM_9K_9t_tech.lef
    read_lef OpenROAD/src/tap/test/gf180/gf180mcu_5LM_1TM_9K_9t_sc.lef

    read_verilog test.v
    link_design test

    initialize_floorplan -die_area {0 0 500 500} -core_area {0 0 500 500} -site GF018hv5v_green_sc9

    add_global_connection -net VDDD -pin_pattern {^VDDD$} -power
    add_global_connection -net GNDD -pin_pattern {^GNDD$} -ground

    set_voltage_domain -power VDDD -ground GNDD

    define_pdn_grid -name "Core"
    add_pdn_ring -grid "Core" -layers {Metal4 Metal5} -widths 5.0 -spacings 2.0 -core_offsets 4.5 -connect_to_pads

    # Comment to avoid error
    # [ERROR PDN-0106] Width (0.0050 um) specified for layer Metal1 is less than minimum width (0.2300 um).
    # add_pdn_stripe -followpins -layer Metal1 -extend_to_core_ring

    add_pdn_stripe -layer Metal4 -width 0.48 -pitch 56.0 -offset 2.0 -extend_to_core_ring
    add_pdn_stripe -layer Metal5 -width 1.40 -pitch 40.0 -offset 2.0 -extend_to_core_ring
    add_pdn_connect -layers {Metal4 Metal5}

   pdngen

**Reply by titan73** [reply]
I suppose the issue comes from my technology lef which I can't provide.
There are no errors or warnings when the techno lef is loader.

I didn't mention some warning before the removal of floating stripes.

    [WARNING PDN-0110] No via inserted between metal4B and alucap at (477.7600, 61.3000) - (478.2400, 62.7000) on VDDD

Is there some debug channels that can be enable to get more details of the reason?

**Reply by maliberty** [reply]
@gadfort can you make some suggestions?

**Reply by gadfort** [reply]
1. you only need one of these:
```
add_pdn_connect -layers {metal4B alucap}
add_pdn_connect -layers {alucap metal4B}
```
2. there is no add_pdn_connect from metal1 to anything so those would likely be floating at the end.
3. you can try and use the debugging gui, you'll be able to see potential issues in the grid before `pdngen` runs, but it's it was written to a developer debugging tool, so it might not be the easiest to use:
```
pdn::debug_renderer 1
pdn::debug_renderer_update
```
beyond this, it would be extremely difficult to debug without a functional testcase.

**Reply by titan73** [reply]
Oh, For 2), I forgot to mentioned this line is commented only on GF180 for the test case as it causes the mentioned error. It is possibly a bug too (I can open an issue for that). Once commented the stripes are created.
With my techno .lef I don't have to comment this line and I don't have the stripes. Only the ring and the sroute in metal1.

**Reply by titan73** [reply]
I can't disclose my techno and standard cells .lef files due to legal reason. A NDA would be needed. How do you proceed generally on this project for this kind of situation?

**Reply by titan73** [reply]
I tested the debug feature. Everything is shown. I can see the stripes.

**Reply by maliberty** [reply]
So this is resolved?

**Reply by titan73** [reply]
No. I can see the shapes being created with the add_pdn_xxx command with the debug turned on but pdngen still remove the stripes.

**Reply by maliberty** [reply]
Have you fixed the connect stmts as suggested?

**Reply by titan73** [reply]
stmts? what does that mean?
What suggestion? If you mean removing the line "add_pdn_connect -layers {alucap metal4B}". Yeah I did. It was a copy-paste leftover. It generates a error anyway. I removed it.

**Reply by maliberty** [reply]
stmts = statements.  If you have removed it then I don't have another suggestion.  If this data is confidential then you might need to engage support from precisioninno.com so it can be shared under NDA.

**Reply by titan73** [reply]
I asked my hierarchy how to deal with that. Since this is research stuff I don't know how it will end up as open source eda is pretty new to my company. But since we're creating a position for a PhD on that subject, that would probably ease things. In the meantime I'm digging into the code of pdn.

**Reply by gadfort** [reply]
Additional thoughts:
1. Do you have a connect statement from `metal1` to `metal4B/alucap` without it the rings would be floating and end up getting removed.
2. If there is no room to place a via on the ring corners, that would also cause some issues. You may want to check the vias available to see what is available.

**Reply by titan73** [reply]
The power ring was not removed as the vias between metal4B and alucap were created. It was only the vias of the power grid that were not created getting the stripes to be removed.

How do I check what vias are available?
In the techno .lef I have 3 vias rules between metal4B and alucap.

On monday, I'll try to use the power ring width for the stripes also to see if I can get the vias created.

**Reply by titan73** [reply, marked as answer]
I've tested with bigger stripes width and this fixed the issue. Vias were created and thus stripes were not removed.
Thanks for the support.

**Reply by maliberty** [reply]
I'm glad you found a solution
